### PR TITLE
State: Parallelize API requests in abstract functions

### LINKED ISSUE #76: State: add FetchSync option
When requesting a resource from the state, it attempts to fetch it from the store, and if not found there starts a request against Discord's API. This quickly becomes a problem of efficiency for methods such as `Permissions`, which, in theory, could trigger up to 3 API requests consecutively, which costs a lot of time.
To improve speed I propose to add a `FetchSync` field to the `State` struct (this could also be named `FetchAsync`, but I think speed is the preferred behavior in most cases, hence default to async behavior). This means a default state will fetch all field asynchronously, starting a goroutine for each individual request, if there are more than one resources requested within a state method. Only if explicitly set to `true` the old, slower behavior will be applied.

Alternatively, this could be made default, without a way to change it.

If approved, I can hand in a PR for this.

### PR DESCRIPTION
**Pull request by mavolin**
Closes #76.

### CODE REVIEW

File: state/state.go [hunk 1]
```diff
@@ -121,34 +121,49 @@ func (s *State) MemberDisplayName(guildID, userID discord.Snowflake) (string, er
 	return member.Nick, nil
 }
 
-func (s *State) AuthorColor(message *gateway.MessageCreateEvent) discord.Color {
+func (s *State) AuthorColor(message *gateway.MessageCreateEvent) (discord.Color, error) {
 	if !message.GuildID.Valid() {
-		return discord.DefaultMemberColor
+		return 0, nil
```
Comment (diamondburned) [inline state/state.go:126]: Is DefaultMemberColor removed?

Comment (mavolin) [reply state/state.go:126]: It's still used in discord.MemberColor

Comment (diamondburned) [reply state/state.go:126]: But this could return a 0 still, so why not make it return default?

Comment (mavolin) [reply state/state.go:126]: Let me fix that.

**Pull request merged**